apps/arweave/src/ar_replica_2_9.erl

```diff
@@ -0,0 +1,473 @@
+	EntropyPartitionSize = get_entropy_partition_size(),
+    PaddedEndOffset = ar_block:get_chunk_padded_offset(AbsoluteChunkEndOffset),
+    BucketStart = ar_chunk_storage:get_chunk_bucket_start(PaddedEndOffset),
+	BucketStart div EntropyPartitionSize.
+
+%% @doc Return the key used to generate the entropy for the 2.9 replication format.
+%% RewardAddr: The address of the miner that mined the chunk.
+%% AbsoluteEndOffset: The absolute end offset of the chunk.
+%% SubChunkStartOffset: The start offset of the sub-chunk within the chunk. 0 is the first
+%% sub-chunk of the chunk, (?DATA_CHUNK_SIZE - ?COMPOSITE_PACKING_SUB_CHUNK_SIZE) is the
+%% last sub-chunk of the chunk.
+-spec get_entropy_key(
+		RewardAddr :: binary(),
+		AbsoluteEndOffset :: non_neg_integer(),
+		SubChunkStartOffset :: non_neg_integer()
+) -> binary().
+get_entropy_key(RewardAddr, AbsoluteEndOffset, SubChunkStartOffset) ->
+	Partition = get_entropy_partition(AbsoluteEndOffset),
+	%% We use the key to generate a large entropy shared by many chunks.
+	EntropyIndex = get_entropy_index(AbsoluteEndOffset, SubChunkStartOffset),
+	crypto:hash(sha256, << Partition:256, EntropyIndex:256, RewardAddr/binary >>).
+
+%% @doc Return the 2.9 entropy sector size - the largest total size in bytes of the contiguous
+%% area where the 2.9 entropy of every chunk is unique.
+-spec get_sector_size() -> pos_integer().
+get_sector_size() ->
+	get_entropy_count() * ?COMPOSITE_PACKING_SUB_CHUNK_SIZE.
+
+%% @doc Return the 0-based index indicating which area within a 2.9 entropy the
+%% given sub-chunk is mapped to (aka slice index). Sub-chunks of the same chunk are mapped to
+%% different entropies but all use the same slice index.
+-spec get_slice_index(
+		AbsoluteChunkEndOffset :: non_neg_integer()
+) -> non_neg_integer().
+get_slice_index(AbsoluteChunkEndOffset) ->
+	PaddedEndOffset = ar_block:get_chunk_padded_offset(AbsoluteChunkEndOffset),
+    BucketStart = ar_chunk_storage:get_chunk_bucket_start(PaddedEndOffset),
+    SubChunkCount = ?REPLICA_2_9_ENTROPY_SIZE div ?COMPOSITE_PACKING_SUB_CHUNK_SIZE,
+	SectorSize = get_sector_size(),
+	(BucketStart div SectorSize) rem SubChunkCount.
+
+
+%%%===================================================================
+%%% Private functions.
+%%%===================================================================
+
+%% @doc Return the total number of entropies generated per partition
+%% in the 2.9 replication format.
+-spec get_entropy_count() -> non_neg_integer().
+get_entropy_count() ->
+	Count = ?PARTITION_SIZE div ?REPLICA_2_9_ENTROPY_SIZE,
+	false = ?PARTITION_SIZE rem ?REPLICA_2_9_ENTROPY_SIZE == 0,
+	%% Add some extra entropies so that the total count is multiple of 32. There are
+    %% 32 sub-chunks per chunk, and each sub-chunk is enciphered with a different entropy. 
+    %% Ensuring we have a multiple of 32 entropies proves very convenient for chunk-by-chunk
+    %% syncing.
+    %% 
+    %% Some entropies will be slightly underused.
```

Comment:
This note is outdated. Every bit of the generated entropy is used now.

Reply:
Updated! https://github.com/ArweaveTeam/arweave/pull/665/commits/eeff94bd8ea1c2d02d219610dee41970ab2d3f54